File: src/execgo_playground/observability/artifacts.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from ..models import ArtifactManifest
from ..utils import ensure_dir, write_json, write_text


class ArtifactManager:
    def __init__(self, root_dir: str | Path, run_id: str) -> None:
        self.root_dir = ensure_dir(root_dir)
        self.run_dir = ensure_dir(Path(root_dir) / run_id)

    def path(self, name: str) -> Path:
        names = {
            "plan": "plan.json",
            "timeline": "timeline.jsonl",
            "snapshots": "execgo_snapshots.jsonl",
            "trace": "adapter_trace.json",
            "result": "result.json",
            "summary": "summary.md",
        }
        return self.run_dir / names[name]

    def write_json(self, name: str, data: Any) -> Path:
        return write_json(self.path(name), data)

    def write_text(self, name: str, content: str) -> Path:
        return write_text(self.path(name), content)

    def manifest(self) -> ArtifactManifest:
        return ArtifactManifest(
            run_dir=str(self.run_dir),
            plan_path=str(self.path("plan")),
            timeline_path=str(self.path("timeline")),
            snapshot_path=str(self.path("snapshots")),
            adapter_trace_path=str(self.path("trace")),
            result_path=str(self.path("result")),
            summary_path=str(self.path("summary")),
        )

File: src/execgo_playground/observability/artifacts.ts
import path from "node:path";
import type { ArtifactManifest } from "../models.js";
import { ensureDir, writeJson, writeText } from "../utils.js";

export class ArtifactManager {
  rootDir: string;
  runDir: string;

  constructor(rootDir: string, runId: string) {
    this.rootDir = rootDir;
    this.runDir = path.join(rootDir, runId);
  }

  async init(): Promise<this> {
    await ensureDir(this.rootDir);
    await ensureDir(this.runDir);
    return this;
  }

  path(name: string): string {
    const names: Record<string, string> = {
      plan: "plan.json",
      timeline: "timeline.jsonl",
      snapshots: "execgo_snapshots.jsonl",
      trace: "adapter_trace.json",
      result: "result.json",
      summary: "summary.md",
    };
    return path.join(this.runDir, names[name]);
  }

  writeJson(name: string, data: unknown): Promise<string> {
    return writeJson(this.path(name), data);
  }

  writeText(name: string, content: string): Promise<string> {
    return writeText(this.path(name), content);
  }

  manifest(): ArtifactManifest {
    return {
      run_dir: this.runDir,
      plan_path: this.path("plan"),
      timeline_path: this.path("timeline"),
      snapshot_path: this.path("snapshots"),
      adapter_trace_path: this.path("trace"),
      result_path: this.path("result"),
      summary_path: this.path("summary"),
    };
  }
}
